test/SIL/memory_lifetime.sil

```diff
@@ -464,6 +464,29 @@ bb0(%0 : @owned $T):
   return %r : $()
 }
 
+// TODO: This should raise a diagnostic:
+// SIL memory lifetime failure in @test_store_to_enum_bad: memory is initialized, but shouldn't
+//
+// There is no diagnostic because the location is marked "trivial" in
+// a persistent way. To fix this, triviality needs to be a data flow
+// bit. Alternatively, we can fix SILGen so that a memory location
```

Comment:
@eeckstein  You're right that we don't need both fixes. The dataflow based check for trivial stores is better, so I can use that. But then that check can't be specific to enum types. It can simply use ownership information instead.

Ownership can either be allowed to propagate, or it can be determined by the type. Allowing it to propagate works well at eliminating useless ARC code. It also simplifies maintaining OSSA form. Ownership propagation is handled well everywhere except the memory lifetime verifier.

The verifier's check for an enum type won't work then because it assumes owership does not propagate. If ownership propagates, then SILGen will generate code like this:
```
sil hidden [ossa] @testTrivialSwitchEnum : $@convention(thin) () -> () {
bb0:
  %1 = enum $Optional<AnyObject>, #Optional.none!enumelt
  switch_enum %1 : $Optional<AnyObject>, case #Optional.some!enumelt: bb1, default bb2

bb1(%3 : $AnyObject):
  %4 = alloc_stack $AnyObject
  store %3 to [trivial] %4 : $*AnyObject
  dealloc_stack %4 : $*AnyObject
  br bb3

bb2:
  br bb3

bb3:
  %40 = tuple ()
  return %40 : $()
}
```